LibJSGCVerifier: False positive in JS::HeapFunction<T>::visit_edges()

LibJSGCVerifier complains that this function doesn't call up to `Base::visit_edges()`:

```
/home/kling/src/serenity/Userland/Libraries/LibJS/Heap/HeapFunction.h:35:5: warning: Missing call to Base::visit_edges
    virtual void visit_edges(Visitor& visitor) override
```

Even though it does!
```cpp
    virtual void visit_edges(Visitor& visitor) override
    {
        Base::visit_edges(visitor);
        visitor.visit_possible_values(m_function.raw_capture_range());
    }
```

cc @mattco98

<!-- POLAR PLEDGE BADGE START -->
## Funding

- We're currently trying [Polar.sh](https://polar.sh/SerenityOS) where you can help fund the work on this issue.
- Contributor(s) receive the funding once the issue is completed & confirmed.

<a href="https://polar.sh/SerenityOS/serenity/issues/23790">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/SerenityOS/serenity/issues/23790/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/SerenityOS/serenity/issues/23790/pledge.svg">
</picture>
</a>
<!-- POLAR PLEDGE BADGE END -->

--- a/Meta/Lagom/Tools/LibJSGCVerifier/src/CellsHandler.cpp
+++ b/Meta/Lagom/Tools/LibJSGCVerifier/src/CellsHandler.cpp
@@ -90,8 +90,9 @@ std::vector<clang::QualType> get_all_qualified_types(clang::QualType const& type
         if (specialization_name == "JS::GCPtr" || specialization_name == "JS::NonnullGCPtr") {
             qualified_types.push_back(type);
         } else {
-            for (size_t i = 0; i < template_specialization->template_arguments().size(); i++) {
-                auto const& template_arg = template_specialization->template_arguments()[i];
+            auto const template_arguments = template_specialization->template_arguments();
+            for (size_t i = 0; i < template_arguments.size(); i++) {
+                auto const& template_arg = template_arguments[i];
                 if (template_arg.getKind() == clang::TemplateArgument::Type) {
                     auto template_qualified_types = get_all_qualified_types(template_arg.getAsType());
                     std::move(template_qualified_types.begin(), template_qualified_types.end(), std::back_inserter(qualified_types));
@@ -143,10 +144,11 @@ FieldValidationResult validate_field(clang::FieldDecl const* field_decl)
             if (template_type_name != "GCPtr" && template_type_name != "NonnullGCPtr")
                 return result;
 
-            if (specialization->template_arguments().size() != 1)
+            auto const template_args = specialization->template_arguments();
+            if (template_args.size() != 1)
                 return result; // Not really valid, but will produce a compilation error anyway
 
-            auto const& type_arg = specialization->template_arguments()[0];
+            auto const& type_arg = template_args[0];
             auto const* record_type = type_arg.getAsType()->getAs<clang::RecordType>();
             if (!record_type)
                 return result;

--- a/Meta/Lagom/Tools/LibJSGCVerifier/src/CellsHandler.cpp
+++ b/Meta/Lagom/Tools/LibJSGCVerifier/src/CellsHandler.cpp
@@ -176,7 +176,8 @@ void CollectCellsHandler::run(clang::ast_matchers::MatchFinder::MatchResult cons
 
     // Cell triggers a bunch of warnings for its empty visit_edges implementation, but
     // it doesn't have any members anyways so it's fine to just ignore.
-    if (record->getQualifiedNameAsString() == "JS::Cell")
+    auto qualified_name = record->getQualifiedNameAsString();
+    if (qualified_name == "JS::Cell")
         return;
 
     auto& diag_engine = result.Context->getDiagnostics();
@@ -216,11 +217,13 @@ void CollectCellsHandler::run(clang::ast_matchers::MatchFinder::MatchResult cons
     // ensuring the classes use JS_CELL/JS_OBJECT, as Base will not be defined if they do not.
 
     MatchFinder base_visit_edges_finder;
-    SimpleCollectMatchesCallback<clang::CXXMemberCallExpr> base_visit_edges_callback("member-call");
+    SimpleCollectMatchesCallback<clang::MemberExpr> base_visit_edges_callback("member-call");
 
-    auto base_visit_edges_matcher = cxxMemberCallExpr(
-        callee(memberExpr(member(hasName("visit_edges")))))
-                                        .bind("member-call");
+    auto base_visit_edges_matcher = cxxMethodDecl(
+        ofClass(hasName(qualified_name)),
+        functionDecl(hasName("visit_edges")),
+        isOverride(),
+        hasDescendant(memberExpr(member(hasName("visit_edges"))).bind("member-call")));
 
     base_visit_edges_finder.addMatcher(base_visit_edges_matcher, &base_visit_edges_callback);
     base_visit_edges_finder.matchAST(*result.Context);

--- a/Userland/Libraries/LibWeb/Fetch/Infrastructure/FetchAlgorithms.cpp
+++ b/Userland/Libraries/LibWeb/Fetch/Infrastructure/FetchAlgorithms.cpp
@@ -47,6 +47,7 @@ FetchAlgorithms::FetchAlgorithms(
 
 void FetchAlgorithms::visit_edges(JS::Cell::Visitor& visitor)
 {
+    Base::visit_edges(visitor);
     visitor.visit(m_process_request_body_chunk_length);
     visitor.visit(m_process_request_end_of_body);
     visitor.visit(m_process_early_hints_response);

--- a/Userland/Libraries/LibWeb/HTML/Navigable.cpp
+++ b/Userland/Libraries/LibWeb/HTML/Navigable.cpp
@@ -57,6 +57,7 @@ public:
 
     virtual void visit_edges(Cell::Visitor& visitor) override
     {
+        Base::visit_edges(visitor);
         visitor.visit(m_response);
     }
 

